File: day15/sol.py
from timing_util import Timing


def get_data(data: str) -> list[str]:
    return data.split(",")


def part1(data: list[str]) -> int:
    score = 0
    for code in data:
        s = 0
        for c in code:
            s += ord(c)
            s *= 17
            s %= 256
        score += s
    return score


def part2(data: list[str]) -> int:
    boxes = [[] for _ in range(256)]
    for code in data:
        i, h = 0, 0
        while code[i].isalpha():
            h += ord(code[i])
            h *= 17
            h %= 256
            i += 1
        lens_id = code[:i]

        box = boxes[h]
        match code[i]:
            case "-":
                for j, (box_lens_id, _) in enumerate(box):
                    if box_lens_id == lens_id:
                        box.pop(j)
                        break
            case "=":
                focal_length = int(code[i + 1])
                for j, (box_lens_id, _) in enumerate(box):
                    if box_lens_id == lens_id:
                        box[j] = (lens_id, focal_length)
                        break
                else:
                    box.append((lens_id, focal_length))

    score = 0
    for box_score, box in enumerate(boxes, 1):
        for slot_score, (_, focal_score) in enumerate(box, 1):
            score += box_score * slot_score * focal_score
    return score


if __name__ == "__main__":
    with Timing("Elapsed time: "):
        with open("inputs/day15.txt") as in_file:
            data = get_data(in_file.read().strip())

        print(f"part1: {part1(data)}")
        print(f"part2: {part2(data)}")

File: day15/sol.go
package main

import (
	"fmt"
	"os"
	"strings"
	"time"
)

type Lens struct {
	name         string
	local_length int
}

func ReadFile(filename string) string {
	bytes, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}
	return strings.TrimSpace(string(bytes))
}

func ParseInput(data string) []string {
	return strings.Split(data, ",")
}

func Part1(data []string) int {
	score := 0
	for _, code := range data {
		s := 0
		for _, c := range code {
			s += int(c)
			s *= 17
			s %= 256
		}
		score += s
	}
	return score
}

func IsAlpha(b byte) bool {
	return ('a' <= b && b <= 'z') || ('A' <= b && b <= 'Z')
}

func Part2(data []string) int {
	boxes := make([][]Lens, 256)
	for _, code := range data {
		i, hash := 0, 0
		for ; i < len(code) && IsAlpha(code[i]); i++ {
			hash += int(code[i])
			hash *= 17
			hash %= 256
		}
		lens_id := code[:i]

		box := boxes[hash]
		switch code[i] {
		case '-':
			for j, lens := range box {
				if lens.name == lens_id {
					boxes[hash] = append(box[:j], box[j+1:]...)
					break
				}
			}
		case '=':
			focal_length := int(code[i+1] - '0')
			replaced := false
			for j, lens := range box {
				if lens.name == lens_id {
					box[j].local_length = focal_length
					replaced = true
					break
				}
			}
			if !replaced {
				boxes[hash] = append(box, Lens{lens_id, focal_length})
			}
		}
	}

	score := 0
	for box_score, box := range boxes {
		for slot_score, lens := range box {
			score += (box_score + 1) * (slot_score + 1) * lens.local_length
		}
	}
	return score
}

func PrintTime(start time.Time, msg string) {
	dur := float64(time.Since(start).Nanoseconds())
	switch {
	case dur < 1e3:
		fmt.Printf("%v %.3fns\n", msg, dur)
	case dur < 1e6:
		fmt.Printf("%v %.3fµs\n", msg, dur/1e3)
	case dur < 1e9:
		fmt.Printf("%v %.3fms\n", msg, dur/1e6)
	default:
		fmt.Printf("%v %.3fs\n", msg, dur/1e9)
	}
}

func main() {
	start := time.Now()
	defer PrintTime(start, "Elapsed time:")

	text := ReadFile("inputs/day15.txt")
	data := ParseInput(text)

	fmt.Printf("Part1: %d\n", Part1(data))
	fmt.Printf("Part2: %d\n", Part2(data))
}
